### PR TITLE
#1029 Don't show an error when explicitly ignoring read-only properties

### LINKED ISSUE #1029: Mapper should not throw an error when explicitly ignoring read-only properties
screen shot 2017-01-03 at 2 35 43 pm

Ran the command `./gradlew assemble` and the annotation processor threw these warnings. Andreas Gudian suggested that they should not do so.

The unmapped properties are generated via the entities' constructors and are marked with `@Transient.` They do not contain accessor methods.

Thanks for reading.

------
Edit by @agudian to clarify what problem we're actually trying to fix here:

Read-Only properties are already ignored automatically and don't show up as unmapped target properties. But explicitly ignoring a read-only property using `@Mapping(target=.., ignore=true)` leads to the error "property X has no write accessor" and supresses any warnings about further unmapped target properties.

### PR DESCRIPTION
**Pull request by agudian**
Fixes #1029.

### CODE REVIEW

**Comment by agudian** [comment]
PRB fails due to outdated license header template - needs #1033 to be merged first.

**Comment by filiphr** [comment]
It looks OK to me

**Comment by filiphr** [comment]
>PRB fails due to outdated license header template - needs #1033 to be merged first.

No you need to update `IgnorePropertyTest.shouldGiveWringingOnUnmapped` as now we do not fail the build if a property with a read accessor is ignored. To fail the build the property should not be ignored.

**Comment by agudian** [comment]
> No you need to update IgnorePropertyTest.shouldGiveWringingOnUnmapped as now we do not fail the compilation if a property with a read accessor is ignored. To fail the build the property should not be ignored.

Yea, that as well. Hmm. It was introduced as part of #833 - but I still think that we should not error out there, but just happily ignore it if the user want's to ignore it... 😁 - so I'll remove that test.

@sjaakd, please shout if I'm missing something here (when you're back, that is).

**Comment by filiphr** [comment]
>Yea, that as well. Hmm. It was introduced as part of #833 - but I still think that we should not error out there, but just happily ignore it if the user want's to ignore it... 😁 - so I'll remove that test.

I agree with you. I think that we should error out if the user does not ignore the property though. We don't have a test where a user targets some property (not ignoring it) without write accessor.

Something like `@Mapping(target = "outdated", constant="true")`

**Comment by agudian** [comment]
> Something like `@Mapping(target = "outdated", constant="true")`

Good idea! I've changed the old test to do just that.

File: processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
     ErroneousTargetHasNoWriteAccessorMapper INSTANCE =
         Mappers.getMapper( ErroneousTargetHasNoWriteAccessorMapper.class );
 
-    @Mapping( target = "hasTallons", ignore = true )
+    @Mapping(target = "hasTallons", constant = "true")
```
Comment (gunnarmorling) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java]: [WTH](http://www.urbandictionary.com/define.php?term=tallon)?!

**Comment by gunnarmorling** [comment]
Reading the title of #833, isn't this exactly the same as this one?

File: processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java (hunk 1 above)
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java]: Dutch humor? 😁

Comment (gunnarmorling) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java]: I'm shocked ;)

**Comment by agudian** [comment]
> Reading the title of #833, isn't this exactly the same as this one?

Kinda. #833 introduced the error message that is shown here. The message is fine if a user tries to configure the mapping for a read-only target. But we missed that ignoring a read-only target should not lead to an error and suppressing the rest of the unmapped properties warnings.
So for this particular case here, we think it's better to leave the user alone and don't bother with a warning or error for a property that we're already ignoring automatically, even if the user explicitly adds the `ignore=true` for that read-only property.

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java [hunk 2]
```diff
@@ -255,17 +255,19 @@ private boolean handleDefinedSourceMappings() {
...
                             method.getResultType().getPropertyReadAccessors().containsKey( mapping.getTargetName() );
-                        ctx.getMessager().printMessage( method.getExecutable(),
-                            mapping.getMirror(),
-                            mapping.getSourceAnnotationValue(),
-                            hasReadAccessor ? Message.BEANMAPPING_PROPERTY_HAS_NO_WRITE_ACCESSOR_IN_RESULTTYPE :
-                                Message.BEANMAPPING_UNKNOWN_PROPERTY_IN_RESULTTYPE,
-                            mapping.getTargetName()
-                        );
-                        errorOccurred = true;
+                        if ( !hasReadAccessor || !mapping.isIgnored() ) {
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java]: Can we de-compose this into two separate ifs? Together with the second one on `hasReadAccessor` inline I find i a bit tough to digest.

**Comment by gunnarmorling** [comment]
Yes, that makes sense. If a target property is read-only, setting `ignore=true` shoudn't raise any warning or error. Any other explicit configuration for that property should result in an error.

**Comment by agudian** [comment]
> Can we de-compose this into two separate ifs? Together with the second one on hasReadAccessor inline I find i a bit tough to digest.

I added a commit for that...

**Comment by sjaakd** [comment]
Hmm,  and I thought I knew my synonyms! 😄. In my world it used to be the claw of a bird of prey. I'm getting old !! 😉

Verstuurd vanaf mijn iPhone

> Op 5 jan. 2017 om 22:35 heeft Gunnar Morling <notifications@github.com> het volgende geschreven:
> 
> I'm shocked ;)

**Comment by filiphr** [comment]
The old test is OK. If you ask me you can merge it

File: processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java (hunk 1 above)
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java]: Me too :smile:. For the record, it should be [talon](https://en.wikipedia.org/wiki/Claw). With one 'l'. Although, that would not get me home-free in the urban dictionary either 😄

Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousTargetHasNoWriteAccessorMapper.java]: I'll change it to something less flowery... 🦄

**Comment by agudian** [comment]
I've squashed / pushed a new commit with reworded properties in the test... 😁

**Pull request merged**